src/libraries/System.Text.Json/gen/Helpers/DiagnosticInfo.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using System.Numerics.Hashing;
+using Microsoft.CodeAnalysis;
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
+    /// Descriptor for diagnostic instances using structural equality comparison.
+    /// Provides a work-around for https://github.com/dotnet/roslyn/issues/68291.
+    /// </summary>
+    public readonly struct DiagnosticInfo : IEquatable<DiagnosticInfo>
+    {
+        public required DiagnosticDescriptor Descriptor { get; init; }
+        public required object?[] MessageArgs { get; init; }
+        public required Location? Location { get; init; }
+
+        public Diagnostic CreateDiagnostic()
+            => Diagnostic.Create(Descriptor, Location, MessageArgs);
+
+        public override readonly bool Equals(object? obj) => obj is DiagnosticInfo info && Equals(info);
+        public readonly bool Equals(DiagnosticInfo other)
+        {
+            return Descriptor.Equals(other.Descriptor) &&
+                MessageArgs.SequenceEqual(other.MessageArgs) &&
```

Comment:
You've implemented equality manually rather than using a record struct in order to handle MessageArgs well?

Reply:
Correct.